ksql-engine/src/main/java/io/confluent/ksql/function/UdfCompiler.java

```diff
@@ -81,22 +123,177 @@
+    try (final InputStream inputStream = getClass().getClassLoader()
+        .getResourceAsStream("KsqlAggregateFunctionTemplate.java")) {
+      final Scanner scanner = new Scanner(inputStream, StandardCharsets.UTF_8.name());
+      final StringBuilder builder = new StringBuilder();
+      while (scanner.hasNextLine()) {
+        builder.append(scanner.nextLine()).append("\n");
+      }
+      udafTemplate = builder.toString();
+    } catch (final IOException io) {
+      throw new KsqlException("Couldn't load UDAF template", io);
+    }
+  }
 
   UdfInvoker compile(final Method method, final ClassLoader loader) {
     try {
-      final IScriptEvaluator scriptEvaluator = createScriptEvaluator(method, loader);
+      final IScriptEvaluator scriptEvaluator = createScriptEvaluator(method,
+          loader,
+          method.getDeclaringClass().getName());
       final String code = generateCode(method);
       return (UdfInvoker) scriptEvaluator.createFastEvaluator(code,
           UdfInvoker.class, new String[]{"thiz", "args"});
     } catch (final KsqlException e) {
       throw e;
-    } catch (Exception e) {
+    } catch (final Exception e) {
       throw new KsqlException("Failed to compile udf wrapper class for "
           + method, e);
     }
   }
 
+  @SuppressWarnings("unchecked")
+  KsqlAggregateFunction compileAggregate(final Method method,
+                                         final ClassLoader loader,
+                                         final String functionName) {
+    final Pair<Type, Type> valueAndAggregateTypes
+        = getValueAndAggregateTypes(method, functionName);
+    try {
+      final String generatedClassName
+          = method.getDeclaringClass().getSimpleName() + "_" + method.getName() + "_Aggregate";
+      final String udafClass = generateUdafClass(generatedClassName, method, functionName);
+      LOGGER.debug("Generated class for functionName={}, method={} class{}\n",
+          functionName,
+          method.getName(),
+          udafClass);
+      final ClassLoader javaSourceClassLoader
+          = createJavaSourceClassLoader(loader, generatedClassName, udafClass);
+
+      final IScriptEvaluator scriptEvaluator =
+          createScriptEvaluator(method,
+              javaSourceClassLoader,
+              UDAF_PACKAGE + generatedClassName);
+      final UdfArgSupplier evaluator = (UdfArgSupplier)
+          scriptEvaluator.createFastEvaluator("return new " + generatedClassName
+                  + "(args, returnType, metrics);",
+              UdfArgSupplier.class, new String[]{"args", "returnType", "metrics"});
+      return evaluator.apply(Collections.singletonList(
+          SchemaUtil.getSchemaFromType(valueAndAggregateTypes.left)),
+          SchemaUtil.getSchemaFromType(valueAndAggregateTypes.right), metrics);
+    } catch (final Exception e) {
+      throw new KsqlException("Failed to compile KSqlAggregateFunction for method='"
```

Comment:
I assume this is the error which will be raised if the argument types don't match the annotation types? Is there an example of the exception raised in this case? Is it intelligible and actionable?

Reply:
I removed the types from the annotation as they aren't required. This exception could be caused by reflection issues etc. It is more of a generic catch all as type checking etc has been handled else where and a produce a descriptive message